rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,555 @@
+# Feature Name: `editor_data_model`
+
+## Summary
+
+This RFC describes the data model and API that the Editor uses in memory, and the basis for communicating with the Game process and for serializing that data to disk (although the serialization format to disk is outside of the scope of this RFC). The data model is largely inspired by the one of [Our Machinery](https://ourmachinery.com/), as described in [this blog post](https://ourmachinery.com/post/the-story-behind-the-truth-designing-a-data-model/).
+
+## Glossary
+
+In this RFC, the following terms are used:
+
+- **Game** refers to any application made with Bevy, which encapsulates proper games but also any other kind of Bevy application (CAD, architecture software, _etc._).
+- **Author** refers to any person (including non-technical) building the Game with the intent of shipping it to end users.
+- **Editor** refers to the software used to edit Game data and prepare them for runtime consumption, and which this RFC describes the data model of.
+- **Runtime** refers to the Game or Editor process running time, as opposed to compiling or building time. When not otherwise specified, this refers to the Game's runtime. In general it's synonymouns of a process running, and is opposed to _build time_ or _compile time_.
+
+## Motivation
+
+We need a data model for the Bevy Editor. We want to make it decoupled from the runtime data model that the Game uses, which is based on compiled Rust types (`World`, _etc._) with a defined native layout from a fixed Rust compiler version (as most types don't use `#[repr(C)]`), so we can handle data migration across Rust compiler versions and therefore across both Editor and Game versions. We also want to make it _transportable_, that is serializable with the intent to transport it over any IPC mechanism, so that the Editor process can dialog with the Game process to enable real-time Game editing ("Play Mode").
+
+## User-facing explanation
+
+The _data model_ of the Editor refers to the way the Editor manipulates editing data, their representation in memory while the Editor executable is running, and the API to do those manipulations. The Editor at its core defines a centralized data model, and all systems use the same API to access that data model, making it the unique source of truth while the Editor process is running. This unicity prevents desynchronization between various subsystems, since they share a same view of the editing data. It also enables global support for common operations such as copy/paste and undo/redo without requiring each subsystem to implement their own variant.

```

Comment:
> What are the systems here? From the context it doesnt look like they are ECS systems. They look more like the editor Extensions you defined in the editor vision.

They're ECS systems, either Editor built-in or from Extensions, I don't think it matters here. I'd avoid talking about Extensions here to avoid getting into details never discussed/RFC'd and partially unrelated.

> This needs to be justified.

Luckily, Our Machinery did that for us already ;) More seriously, what kind of justification do you want to hear here? Doesn't this directly derives from the fact there's a single source of truth, so copy/paste and undo/redo can affect that unique source, and the effect be observed by all other systems?

Reply:
> What kind of justification do you want to hear here?

Yes, what you exposed is what I needed to hear 🙂. Maybe it should be made more clear, or I might just have to reread that paragraph...

Reply:
We're in the user-facing part though, do we need to add a justification that this is technically feasible there?

Suggestion:
> "It also enables global support for common operations such as copy/paste and undo/redo, since all ECS systems will observe the same copy/paste/undo/redo changes to that unique source of truth."

Reply:
I added that suggestion in latest version.